src/Data/Variant.purs

```diff
@@ -217,6 +223,211 @@ instance eqVariant ∷ (R.RowToList r rl, VariantTags rl, VariantEqs rl) ⇒ Eq
     in
       lookupEq tags eqs c1 c2
 
+class VariantBounded (rl ∷ R.RowList) where
+  variantBounded ∷ RLProxy rl → L.List (BoundedDict VariantCase)
+
+instance boundedVariantNil ∷ VariantBounded R.Nil where
+  variantBounded _ = L.Nil
+
+instance boundedVariantCons ∷ (VariantBounded rs, Bounded a) ⇒ VariantBounded (R.Cons sym a rs) where
+  variantBounded _ = L.Cons dict (variantBounded (RLProxy ∷ RLProxy rs))
+    where
+    dict ∷ BoundedDict VariantCase
+    dict =
+      { top: coerceA top
+      , bottom: coerceA bottom
+      }
+
+    coerceA ∷ a → VariantCase
+    coerceA = unsafeCoerce
+
+instance boundedVariant
+  ∷ ( R.RowToList r rl
+    , VariantTags rl
+    , VariantEqs rl
+    , VariantOrds rl
+    , VariantBounded rl
+    )
+  ⇒ Bounded (Variant r)
+  where
+    top =
+      let
+        prx = RLProxy ∷ RLProxy rl
+        tags = variantTags prx
+        dicts = variantBounded prx
+        zipped = L.zip tags dicts
+        hd = fromJust "top" $ L.head $ L.sortBy (\a b → compare (T.fst b) (T.fst a)) zipped
+      in unsafeCoerce { type: T.fst hd, value: (T.snd hd).top }
```

Comment:
I don't think there's a reason to `zip` and `sortBy` here. `RowToList` is already sorted, so this should just be a matter of using `head` and `last`. Can we have an annotated `VariantRep a -> Variant r` coercion?

Reply:
I know it's more code, but I'd like to avoid making extra traversals/allocations over the lists in general. Since this is typeclass dispatch it needs to be as efficient as possible, even if that means writing special purpose loops over lists.